### PR TITLE
Fixes #3464 where http requests don't honor cancellation

### LINKED ISSUE #3464: HttpClient isn't always canceling in response to cancellation requests
We first found this issue with a WCF negative test that expected a TimeOutException when setting a ‘SendTimeout’ on our BasicHttpBinding and calling a service operation that took longer to return.

Previously this test passed with the elapsed time being just a little longer than the ‘SendTimeout’ value. As of a couple weeks ago we still got the TimeOutException but the elapsed time seemed to correlate to however long it took for the call to return from the service operation.

So if the ‘SendTimeout’ was set to 5 milliseconds and the service operation waited 30 milliseconds to return, the elapsed time would be something over 30 milliseconds.
As best we could determine it has something to do with the CancellationTokenSource perhaps when it is creating the linked token source and passing it to the SetTimeout method. (HttpClient.cs SendAsync method)

We isolated this issue using no WCF code with the following client side code…
```csharp
        Stopwatch watch = new Stopwatch();
        HttpClient client = new HttpClient();
        HttpRequestMessage message = new HttpRequestMessage();
        message.RequestUri = new Uri("http://localhost:54464/Home/About");
        var cts = new CancellationTokenSource(TimeSpan.FromSeconds(10));
        try
        {
            watch.Start();
            var responseTask =  client.SendAsync(message, cts.Token);
            responseTask.Wait();            
        }
        catch (Exception ex)
        {
            Debug.WriteLine(ex.Message + "\n" + ex.StackTrace);
        }
        finally{
            watch.Stop();
        }

        Assert.InRange<long>(watch.ElapsedMilliseconds, 8000, 15000);
```
In this example we are just calling the ‘About’ page of an ASP.NET MVC application (default VS template) and just added the following in the HomeController.About() method…
```csharp
        public ActionResult About()
        {
            ViewBag.Message = "Your application description page.";
            Thread.Sleep(30*1000);
            return View();
        }
```
**NOTE: I have no idea why but it seems to work as expected when in debug mode, the issue reproduces when executed normally.**

### PR DESCRIPTION
**Pull request by mconnew**
Add a registration for the passed in cancellation token which cancels the CancellationTokenSource.Token that is returned to the caller.

### CODE REVIEW

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs [hunk 1]
```diff
@@ -1084,189 +1084,192 @@ private void SaveServerCredentialsToCache(Uri uri, uint authScheme, ICredentials
                 return;
             }
 
-            try
+            using(state.CancellationToken.Register(() => state.Tcs.TrySetCanceled()))
```
Comment (Maxwe11) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: :grey_exclamation: delegate and closure allocation

    state.CancellationToken.Register(s => ((RequestState)s).Tcs.TrySetCanceled(), state)

**Comment by stephentoub** [comment]
cc: @davidsh, @cipop

**Comment by stephentoub** [comment]
I would recommend not doing this. When cancellation is requested, the returned task will be immediately canceled, which means the caller won't know when the currently outstanding operations have finished.

**Comment by davidsh** [comment]
I made a change in recent PR to remove this exact construct since it was causing handle leaks.

```c#
using(state.CancellationToken.Register(() => state.Tcs.TrySetCanceled()))
```

Right now, the `WinHttpHandler` is sync pattern with WinHTTP.  Cancellation in general is cooperative and best-effort.  This will improve with #3000.  I am working on a large PR in this area and will consider this issue as another test case for async cancellation handling.

But this PR itself is not safe and will not be considered at this time.

**Comment by mconnew** [comment]
@davidsh, if we changed our usage to timeout the request outside of HttpClient, and then called Dispose() on the HttpRequestMessage once our own timeout expires, would we still see a handle leak?

**Comment by davidsh** [comment]
@mconnew .  Disposing the HttpRequestMessage does not affect any async operations happening on that request message.  All it would do is call Dispose() on the HttpRequestMessage.Content assuming there is a non-null content.

I don't understand what you mean by this:

>to timeout the request outside of HttpClient, and then called Dispose() on the HttpRequestMessage once our own timeout expires,

In general, trying to do the code in this PR only artificially cancels the awaited task.  Leaving the async operation still running wild only leads to problem.  

We will be looking at trying to make cancellation as responsive as possible as we continue our work on WinHttpHandler.

**Comment by stephentoub** [comment]
In the meantime, since it's very unlikely this is the form a solution would take and since there's a separate set of issues around this, I'd recommend closing the PR.

**Comment by mconnew** [comment]
@davidsh, how is this any different than doing the following:
```csharp
            var tcs = new TaskCompletionSource<bool>();
            var cts = new CancellationTokenSource(TimeSpan.FromMilliseconds(500));
            var reqTask = httpClient.SendAsync(reqMsg);
            using (cts.Token.Register(() => tcs.TrySetCanceled()))
            {
                await Task.WhenAny(tcs.Task, reqTask);
            }
            if (!reqTask.IsCompleted)
            {
                reqMsg.Dispose();
            }
```
In both cases the request still continues on, and in the WCF use case we call Dispose() to clean up.

**Comment by davidsh** [comment]
If you want the HttpClient APIs to observe your cancellation request, you need to pass in the cancellation token into this API:

```c#
var reqTask = httpClient.SendAsync(reqMsg);
```

Should be:

```c#
var reqTask = httpClient.SendAsync(reqMsg, cts.Token);
```

And even then, HttpClient will try to respond to your cancellation request on a best-effort basis.

**Comment by mconnew** [comment]
That's what we're doing, And currently best-effort basis appears to not include the time while waiting for the response headers to come back. This could end up being a very long time. Can you please clarify, is it safe to call .Dispose() on an HttpRequestMessage while the request is still outstanding?

**Comment by davidsh** [comment]
>And currently best-effort basis appears to not include the time while waiting for the response headers to come back.

This is because WinHttpHandler uses SYNC pattern WinHTTP APIs.  So, therefore, it is not looking at the cancellation token until WinHTTP API calls return back to the caller (WinHttpHandler).  When this handler is converted to async pattern, it will be more responsive.

In terms of this:

>Can you please clarify, is it safe to call .Dispose() on an HttpRequestMessage while the request is still outstanding?

I would say that it is a NO-OP for cancelling the async operation. It is also NOT SAFE if you are POST'ing content because you are Dispos'ing it while an async operation is in flight.  Depending on race conditions, it might be ok (if they content was read already), or you'll generate some error from failing to read the content to POST.

**Comment by davidsh** [comment]
Closing this PR to prevent it from being accidentally merged.

**Comment by mconnew** [comment]
As the example code I provided in comment does the same thing as the PR (that is abandon the in-progress request) and calls Dispose() the same as WCF was using HttpClient, I don't understand why the example code is safe but the PR isn't?

**Pull request closed without merging**